jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java

```diff
@@ -197,242 +189,92 @@
-                        }
-                }
-            }
-            if (buf == null)
-                return null;
+            char c = path.charAt(i);
+            if (c > 0x7F) // 8-bit +
+                needsByteEncoding = true;
+            else if (ENCODE_PATH_NEEDS_ENCODING[c])
+                needsEncoding = true;
         }
 
-        int i;
+        if (needsByteEncoding)
+            return encodePathBytes(path);
+        else if (needsEncoding)
+        {
+            return encodePathString(path);
+        }
+        else
+            return path;
+    }
 
-        loop:
-        for (i = offset; i < path.length(); i++)
+    private static String encodePathString(String path)
+    {
+        StringBuilder buf = new StringBuilder(path.length() * 2);
+        int length = path.length();
+        for (int i = 0; i < length; i++)
         {
             char c = path.charAt(i);
-            switch (c)
+            if (ENCODE_PATH_NEEDS_ENCODING[c])
             {
-                case '%':
-                    buf.append("%25");
-                    continue;
-                case '?':
-                    buf.append("%3F");
-                    continue;
-                case ';':
-                    buf.append("%3B");
-                    continue;
-                case '#':
-                    buf.append("%23");
-                    continue;
-                case '"':
-                    buf.append("%22");
-                    continue;
-                case '\'':
-                    buf.append("%27");
-                    continue;
-                case '<':
-                    buf.append("%3C");
-                    continue;
-                case '>':
-                    buf.append("%3E");
-                    continue;
-                case ' ':
-                    buf.append("%20");
-                    continue;
-                case '[':
-                    buf.append("%5B");
-                    continue;
-                case '\\':
-                    buf.append("%5C");
-                    continue;
-                case ']':
-                    buf.append("%5D");
-                    continue;
-                case '^':
-                    buf.append("%5E");
-                    continue;
-                case '`':
-                    buf.append("%60");
-                    continue;
-                case '{':
-                    buf.append("%7B");
-                    continue;
-                case '|':
-                    buf.append("%7C");
-                    continue;
-                case '}':
-                    buf.append("%7D");
-                    continue;
-
-                default:
-                    if (c < 0x20 || c >= 0x7f)
-                    {
-                        bytes = path.getBytes(URIUtil.__CHARSET);
-                        break loop;
-                    }
-                    buf.append(c);
+                buf.append('%');
+                TypeUtil.toHex(c, buf);
```

Comment:
You can safely cast `c` to a `byte` here, since this method is called only if all chars are ASCII, therefore bytes.